### ISSUE TITLE
API Documentation

### ISSUE THREAD

**Issue by malb** [fplll/fplll#235]
We should use Doxygen's

- [examples](http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdexample) to give easily digestible examples of how to use the library 

- [modules](http://www.stack.nl/~dimitri/doxygen/manual/grouping.html) to distinguish high-level and low-level functionality

**Comment by Vaggelis1994** [comment]
Hi Martin, 

considering the documentation, I would like to know how could it be possible to make a contribution. I have spent several time on the API, mostly what regards BKZ, and I would be really happy, if I could help.

**Comment by malb** [comment]
Great! You could take a look at the Doxygen documentation http://www.stack.nl/~dimitri/doxygen/ (and the links above) to get an idea about the format for the code. You can see `modules` in action, e.g. here https://malb.bitbucket.io/gghlite-flint/index.html Then, I'd say, it's mainly a task of going through the relevant header files and checking 

- Is every function documentation: what does it do, what are its parameters, what are its assumptions, what are its outputs and side effects
- Does it make sense to tag with with an example.
- We could have modules such as "user" and "developer" to tag functions as highlevel and lowlevel.

**Comment by Vaggelis1994** [comment]
Hi again Martin,

firstly, I am sorry for my late response. I would like to know whether there is still need for the automation of the documentation. 

If so, I would like to know which are the cases of using an example tag. Could you might providing me an example? Furthermore, concerning the "user" and "developer" tag functions, if I have understood it properly, for example, the `bkz_reduction` function would be tagged with a "developer" tag and `BKZReduction` with a "user" tag? 
 
Finally, I made some tests with doxygen documentation and I am familiarized now with the way it works.

**Comment by malb** [comment]
Hi there, not sure what you mean by automating the documentation. We're creating it semi-automagically at https://fplll.github.io/fplll/ Improvements to the documentation would be very welcome! Wouldn't the tags be exactly the other way around: `bkz_reduction` is a high-level fire-and-forget function whereas `BKZReduction` requires to create a chain of objects to be used but offers greater flexibility. I'd assume there are only very few functions that get a `user` or `highlevel` tag e.g. `bkz_reduction` and `lll_reduction`.

**Comment by Vaggelis1994** [comment]
Yeah, you are absolutely right. This is what I meant. Concerning examples, an example tag would be used for an example such as `./latticegen r 10 1000 | ./fplll` , as stated at [https://fplll.github.io/fplll/](https://fplll.github.io/fplll/), or is it something different?

**Comment by malb** [comment]
Heya, take a look at e.g. https://bitbucket.org/malb/m4ri/src/a1af290b8e9757142daad5750bc2181be53d0d93/m4ri/brilliantrussian.h?at=master&fileviewer=file-view-default#brilliantrussian.h-207 The tag `\example` there points to a file where the function is called to give the reader some guidance of how it's used. It then shows up at https://malb.bitbucket.io/m4ri/examples.html See also https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdexample
